src/python/pants/backend/python/macros/deprecation_fixers.py

```diff
@@ -0,0 +1,223 @@
+from pants.core.goals.update_build_files import (
+    DeprecationFixerRequest,
+    RewrittenBuildFile,
+    RewrittenBuildFileRequest,
+)
+from pants.engine.addresses import Address, Addresses, AddressInput, BuildFileAddress
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    AllTargets,
+    Dependencies,
+    DependenciesRequest,
+    ExplicitlyProvidedDependencies,
+    MultipleSourcesField,
+    UnexpandedTargets,
+)
+from pants.engine.unions import UnionRule
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+from pants.util.strutil import bullet_list
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True, order=True)
+class GeneratorRename:
+    build_path: str
+    alias: str
+    new_name: str | None
+
+
+@dataclass(frozen=True)
+class MacroRenames:
+    generators: tuple[GeneratorRename, ...]
+    # Includes the alias for the macro.
+    generated: FrozenDict[Address, tuple[Address, str]]
+
+
+@rule(desc="Determine how to rename Python macros to target generators", level=LogLevel.DEBUG)
+async def determine_macro_changes(all_targets: AllTargets) -> MacroRenames:
+    # Strategy: Find `python_requirement` targets who depend on a `_python_requirements_file`
+    # target to figure out which macros we have. Note that context-aware object factories (CAOFs)
+    # are not actual targets and are "erased", so this is the way to find the macros.
+    #
+    # We also need to figure out if the new target generator can use the default `name=None` or
+    # if it needs to set an explicit name, based on whether it's the build root and whether the
+    # default is already taken.
+
+    dirs_with_default_name = set()
+    python_requirement_dependencies_fields = set()
+    for tgt in all_targets:
+        if tgt.address.is_default_target:
+            dirs_with_default_name.add(tgt.address.spec_path)
+        if isinstance(tgt, PythonRequirementTarget) and tgt[Dependencies].value is not None:
+            python_requirement_dependencies_fields.add(tgt[Dependencies])
+
+    build_file_addresses_per_tgt = await MultiGet(
+        Get(BuildFileAddress, Address, deps_field.address)
+        for deps_field in python_requirement_dependencies_fields
+    )
+    explicit_deps_per_tgt = await MultiGet(
+        Get(ExplicitlyProvidedDependencies, DependenciesRequest(deps_field))
+        for deps_field in python_requirement_dependencies_fields
+    )
+    deps_per_tgt = await MultiGet(
+        Get(UnexpandedTargets, Addresses(explicit_deps.includes))
+        for explicit_deps in explicit_deps_per_tgt
+    )
+
+    generators = set()
+    generated = {}
+    for python_req_deps_field, build_file_addr, deps in zip(
+        python_requirement_dependencies_fields, build_file_addresses_per_tgt, deps_per_tgt
+    ):
+        generator_tgt = next((tgt for tgt in deps if isinstance(tgt, PythonRequirementsFile)), None)
```

Comment:
Just noting that the `PythonRequirementsFile` is a bit misleading, more so for it's doc string stating it is for `requirements.txt` files, when in fact it is also used for `pipenv.lock` and `pyproject.toml` files.